### PR TITLE
net/icmp: icmp bug fix and enhancement

### PR DESCRIPTION
**Pull request by anchao**
## Summary

net/icmp: fix race condition in icmp recvmsg
net/icmp: add nonblocking support
net/icmp: consume the data length to avoid duplicate packet
net/icmp: fix invalid condition comparison

## Impact

ping/ping6

## Testing

ping/ping6 works as expected

### CODE REVIEW

File: net/icmp/icmp_recvmsg.c [hunk 1]
```diff
@@ -162,7 +162,7 @@ static uint16_t recvfrom_eventhandler(FAR struct net_driver_s *dev,
 
           /* Return the size of the returned data */
 
-          DEBUGASSERT(recvsize > INT16_MAX);
+          DEBUGASSERT(recvsize < INT16_MAX);
```
Comment (yamt) [inline net/icmp/icmp_recvmsg.c:165]: maybe "<=" ?

File: net/icmpv6/icmpv6_recvmsg.c [hunk 2]
```diff
@@ -169,7 +169,7 @@ static uint16_t recvfrom_eventhandler(FAR struct net_driver_s *dev,
 
           /* Return the size of the returned data */
 
-          DEBUGASSERT(recvsize > INT16_MAX);
+          DEBUGASSERT(recvsize < INT16_MAX);
```
Comment (yamt) [inline net/icmpv6/icmpv6_recvmsg.c:172]: <= ?

File: net/icmp/icmp_recvmsg.c (hunk 1 above)
Comment (anchao) [reply net/icmp/icmp_recvmsg.c]: Done

File: net/icmpv6/icmpv6_recvmsg.c (hunk 2 above)
Comment (anchao) [reply net/icmpv6/icmpv6_recvmsg.c]: Done

File: net/icmp/icmp_recvmsg.c [hunk 3]
```diff
@@ -389,111 +389,114 @@ ssize_t icmp_recvmsg(FAR struct socket *psock, FAR struct msghdr *msg,
...
     }
 
+  /* Get the device that was used to send the ICMP request. */
+
+  dev = conn->dev;
+  DEBUGASSERT(dev != NULL);
+  if (dev == NULL)
+    {
+      ret = -EPROTO;
+      goto errout;
```
Comment (xiaoxiang781216) [inline net/icmp/icmp_recvmsg.c:399]: net_unlock shouldn't be called here.

File: net/icmp/icmp_recvmsg.c [hunk 4]
```diff
@@ -389,111 +389,114 @@ ssize_t icmp_recvmsg(FAR struct socket *psock, FAR struct msghdr *msg,
       goto errout;
```
Comment (xiaoxiang781216) [inline net/icmp/icmp_recvmsg.c:389]: net_unlock shouldn't be called here.

Comment (anchao) [reply net/icmp/icmp_recvmsg.c]: Done

File: net/icmp/icmp_recvmsg.c [hunk 5]
```diff
@@ -384,108 +385,116 @@ ssize_t icmp_recvmsg(FAR struct socket *psock, FAR struct msghdr *msg,
...
   if (conn->nreqs < 1)
     {
-      ret = -EPROTO;
-      goto errout;
+      return -EPROTO;
     }
 
+  /* Get the device that was used to send the ICMP request. */
+
+  dev = conn->dev;
```
Comment (xiaoxiang781216) [inline net/icmp/icmp_recvmsg.c:393]: should we keep it in the original place?

**Review by xiaoxiang781216 — APPROVED** [review]
LGTM.

**Comment by xiaoxiang781216** [comment]
@yamt @antmerlino could you take a look again?

**Review by yamt — APPROVED** [review]
lgtm

**Comment by antmerlino** [comment]
@xiaoxiang781216 give me a day or so to look at this again

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 give me a day or so to look at this again

Sure.

File: net/icmp/icmp_recvmsg.c [hunk 6]
```diff
@@ -181,7 +181,8 @@ static uint16_t recvfrom_eventhandler(FAR struct net_driver_s *dev,
 
           /* Indicate that the data has been consumed */
 
-          flags &= ~ICMP_NEWDATA;
+          flags     &= ~ICMP_NEWDATA;
+          dev->d_len = 0;
```
[the remark this answers is not in the document]
Comment (anchao) [reply net/icmp/icmp_recvmsg.c:185]: Hi, @antmerlino 

This change follows the logic of other protocol stacks, d_Len should be set to 0 if the buffer is consumed
https://github.com/apache/incubator-nuttx/blob/3859031a6bfdc987b6704f5460762e91e8d84b8e/net/tcp/tcp_recvfrom.c#L215

https://github.com/apache/incubator-nuttx/blob/3859031a6bfdc987b6704f5460762e91e8d84b8e/net/udp/udp_recvfrom.c#L179

[the remark this answers is not in the document]
Comment (antmerlino) [reply net/icmp/icmp_recvmsg.c:185]: @anchao Thanks for clarifying. I didn't see them because they were in a function called from recv_handler.

**Review by antmerlino — APPROVED** [review]
LGTM

**Pull request merged**